# 01. Importing libraries

In [1]:
# Import libraries
import warnings
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt

from pandas.core.common import SettingWithCopyWarning
from scipy.stats import norm

# 02. Importing data

In [2]:
# Define csv path
path = r'C:\Users\lucav\Desktop\Instacart Basket Analysis'

In [3]:
# Import data from 'orders_products_merged.pkl'
warnings.simplefilter(action='ignore', category=FutureWarning) # Added to ignore warnings
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_frequencies.pkl'))

# 03. Task

## I. Calculate the mean of the 'order_number' column grouped by the 'department_id' column

In [4]:
# Create a subset with the first one million rows of 'ords_prods_merge'
df = ords_prods_merge[:1000000]

In [5]:
# In the 'df' subset, calculate the mean of the 'order_number' column grouped by the 'department_id' column
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [6]:
# In the 'ords_prods_merge' dataframe, calculate the mean of the 'order_number' column grouped by the 'department_id' column
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## II. Analyze the result. How do the results for the entire dataframe differ from those of the subset?

The subset contains only the first one million rows of 'ords_prods_merge', thus only partial information is available.
On the other hand, when analysing the entire dataframe, all the information is taken into account.

## III. Create a loyalty flag for existing customers using the transform() and loc() functions

In [7]:
# Create a flag for customers based on the number of orders
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [8]:
# Print the first 100 rows of 'ords_prods_merge'
pd.options.display.max_rows = None
ords_prods_merge.head(100)

order_id  user_id  order_number  day_of_week  hour_of_day  \
0    2539329        1             1            2            8   
1    2398795        1             2            3            7   
2     473747        1             3            3           12   
3    2254736        1             4            4            7   
4     431534        1             5            4           15   
5    3367565        1             6            2            7   
6     550135        1             7            1            9   
7    3108588        1             8            1           14   
8    2295261        1             9            1           16   
9    2550362        1            10            4            8   
10   2968173       15            15            1            9   
11   1870022       15            17            2           16   
12   1911383       15            18            2           11   
13   2715276       15            21            1            9   
14    487368       15            22            1           10   
15   2293453       19             2            5           14   
16   1973799       19             5            6           12   
17    532817       19             7            4           17   
18   1573906       21            10            3           10   
19   1593000       31            10            3            8   
20   2231262       31            17            3           11   
21   2580647       43             6            4           16   
22   2187180       43             9            4           12   
23   2497897       52             1            1            9   
24   1318871       52             2            1           10   
25   1261384       52             3            1           10   
26    580568       52             5            2           10   
27   2428073       52             7            2            7   
28    342306       52             9            1           11   
29   1520353       52            10            2           10   
30   2542086       52            11            2            9   
31    944694       52            12            1           16   
32   1268191       52            15            2           11   
33   1122089       52            17            2           10   
34   1498922       52            18            2           10   
35    180919       52            19            3            8   
36   2695875       52            21            5           10   
37   3205960       67             1            3           11   
38    970922       67             2            1           15   
39    596585       67             3            1           13   
40   2037087       67             4            2            9   
41   2474111       67             5            3           14   
42   1568481       67             6            3           12   
43   2048527       67             8            2           11   
44   2389031       67             9            2           14   
45   2378562       67            10            1           10   
46   1515698       67            11            1           11   
47   1928050       67            12            2           12   
48   1949274       67            13            1           10   
49    551869       67            14            1            9   
50    768572       67            16            4           10   
51    150528       67            18            5           16   
52     29758       67            19            0           14   
53   1558976       67            21            1            9   
54    863974       67            22            1           11   
55   2308399       67            23            1           14   
56   2759128       81             3            1            9   
57   1812985       81             5            2           20   
58   1889835       82             1            3           15   
59    499460       82             3            5           10   
60   2728456       82             4    

## IV. Check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). Do prices of products purchased by loyal customers differ from those purchased by regular or new customers?

In [9]:
# Calculate the mean, min and max of the 'order_number' column grouped by the 'department_id' column
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

The results show that the loyalty category with the highest average spending is "New customer" with an amount of 13.29, followed by "Regular customer" with 12.49 and 'Loyal customer' with 10.38. However, these data do not provide any information on whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [10]:
# Count the values in the 'loyalty_flag' column
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [11]:
# Calculate the number of products sold within each price range for each loyalty category
ords_prods_merge.groupby('loyalty_flag')['price_range_loc'].value_counts(dropna = False)

loyalty_flag      price_range_loc   
Loyal customer    Mid-range product      6949467
                  Low-range product      3214690
                  High-range product      119936
New customer      Mid-range product      4205569
                  Low-range product      1952769
                  High-range product       85652
Regular customer  Mid-range product     10705824
                  Low-range product      4958862
                  High-range product      212090
Name: price_range_loc, dtype: int64

It can be observed that, for each loyalty category, the percentages of products sold within each price range are as follows: 
- Loyal customer:
                - Mid-range product: 67.57%
                - Low-range product: 31.26%
                - High-range product: 1.17%
- New customer:
                - Mid-range product: 67.35%
                - Low-range product: 31.27%
                - High-range product: 1.37%
- Regular customer:
                - Mid-range product: 67.43%
                - Low-range product: 31.23%
                - High-range product: 1.34%

As the percentages suggest, the number of sales of low-range products is consistent across the three categories. However, looking at the other price ranges, loyal customers have recorded the largest percentage of mid-range product sales and the lowest percentage of high-range product sales. Thus, data suggest that **loyal customers tend to purchase more mid-range products** and **less high-range products**, when compared to new and regular customers.

## V. Create a spending flag for each user based on the average price across all their orders

In [12]:
# Create a column 'customer_spending' grouping by 'user_id' and calculating the average price for each user
ords_prods_merge['customer_spending'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

# If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender”
ords_prods_merge.loc[ords_prods_merge['customer_spending'] >= 10, 'spending_flag'] = 'High spender'

# If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender”
ords_prods_merge.loc[ords_prods_merge['customer_spending'] < 10, 'spending_flag'] = 'Low spender'

In [13]:
# Print the first 100 rows of 'ords_prods_merge'
pd.options.display.max_rows = None
ords_prods_merge.head(100)

order_id  user_id  order_number  day_of_week  hour_of_day  \
0    2539329        1             1            2            8   
1    2398795        1             2            3            7   
2     473747        1             3            3           12   
3    2254736        1             4            4            7   
4     431534        1             5            4           15   
5    3367565        1             6            2            7   
6     550135        1             7            1            9   
7    3108588        1             8            1           14   
8    2295261        1             9            1           16   
9    2550362        1            10            4            8   
10   2968173       15            15            1            9   
11   1870022       15            17            2           16   
12   1911383       15            18            2           11   
13   2715276       15            21            1            9   
14    487368       15            22            1           10   
15   2293453       19             2            5           14   
16   1973799       19             5            6           12   
17    532817       19             7            4           17   
18   1573906       21            10            3           10   
19   1593000       31            10            3            8   
20   2231262       31            17            3           11   
21   2580647       43             6            4           16   
22   2187180       43             9            4           12   
23   2497897       52             1            1            9   
24   1318871       52             2            1           10   
25   1261384       52             3            1           10   
26    580568       52             5            2           10   
27   2428073       52             7            2            7   
28    342306       52             9            1           11   
29   1520353       52            10            2           10   
30   2542086       52            11            2            9   
31    944694       52            12            1           16   
32   1268191       52            15            2           11   
33   1122089       52            17            2           10   
34   1498922       52            18            2           10   
35    180919       52            19            3            8   
36   2695875       52            21            5           10   
37   3205960       67             1            3           11   
38    970922       67             2            1           15   
39    596585       67             3            1           13   
40   2037087       67             4            2            9   
41   2474111       67             5            3           14   
42   1568481       67             6            3           12   
43   2048527       67             8            2           11   
44   2389031       67             9            2           14   
45   2378562       67            10            1           10   
46   1515698       67            11            1           11   
47   1928050       67            12            2           12   
48   1949274       67            13            1           10   
49    551869       67            14            1            9   
50    768572       67            16            4           10   
51    150528       67            18            5           16   
52     29758       67            19            0           14   
53   1558976       67            21            1            9   
54    863974       67            22            1           11   
55   2308399       67            23            1           14   
56   2759128       81             3            1            9   
57   1812985       81             5            2           20   
58   1889835       82             1            3           15   
59    499460       82             3            5           10   
60   2728456       82             4    

## VI. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column

In [14]:
# Create a column 'order_frequency' grouping by 'user_id' and calculating the median days since prior order for each user
ords_prods_merge['order_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

# If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer”
ords_prods_merge.loc[ords_prods_merge['order_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

# If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer”
ords_prods_merge.loc[(ords_prods_merge['order_frequency'] <= 20) & (ords_prods_merge['order_frequency'] > 10), 'order_frequency_flag'] = 'Regular customer'

# If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer”
ords_prods_merge.loc[ords_prods_merge['order_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [15]:
# Print the first 100 rows of 'ords_prods_merge'
pd.options.display.max_rows = None
ords_prods_merge.head(100)

order_id  user_id  order_number  day_of_week  hour_of_day  \
0    2539329        1             1            2            8   
1    2398795        1             2            3            7   
2     473747        1             3            3           12   
3    2254736        1             4            4            7   
4     431534        1             5            4           15   
5    3367565        1             6            2            7   
6     550135        1             7            1            9   
7    3108588        1             8            1           14   
8    2295261        1             9            1           16   
9    2550362        1            10            4            8   
10   2968173       15            15            1            9   
11   1870022       15            17            2           16   
12   1911383       15            18            2           11   
13   2715276       15            21            1            9   
14    487368       15            22            1           10   
15   2293453       19             2            5           14   
16   1973799       19             5            6           12   
17    532817       19             7            4           17   
18   1573906       21            10            3           10   
19   1593000       31            10            3            8   
20   2231262       31            17            3           11   
21   2580647       43             6            4           16   
22   2187180       43             9            4           12   
23   2497897       52             1            1            9   
24   1318871       52             2            1           10   
25   1261384       52             3            1           10   
26    580568       52             5            2           10   
27   2428073       52             7            2            7   
28    342306       52             9            1           11   
29   1520353       52            10            2           10   
30   2542086       52            11            2            9   
31    944694       52            12            1           16   
32   1268191       52            15            2           11   
33   1122089       52            17            2           10   
34   1498922       52            18            2           10   
35    180919       52            19            3            8   
36   2695875       52            21            5           10   
37   3205960       67             1            3           11   
38    970922       67             2            1           15   
39    596585       67             3            1           13   
40   2037087       67             4            2            9   
41   2474111       67             5            3           14   
42   1568481       67             6            3           12   
43   2048527       67             8            2           11   
44   2389031       67             9            2           14   
45   2378562       67            10            1           10   
46   1515698       67            11            1           11   
47   1928050       67            12            2           12   
48   1949274       67            13            1           10   
49    551869       67            14            1            9   
50    768572       67            16            4           10   
51    150528       67            18            5           16   
52     29758       67            19            0           14   
53   1558976       67            21            1            9   
54    863974       67            22            1           11   
55   2308399       67            23            1           14   
56   2759128       81             3            1            9   
57   1812985       81             5            2           20   
58   1889835       82             1            3           15   
59    499460       82             3            5           10   
60   2728456       82             4    

## VII. Export 'orders_products_merge' as 'orders_products_merged_frequencies&flags.pkl'

In [16]:
ords_prods_merge.to_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged_frequencies&flags.pkl'))